### DISCUSSION TITLE
Questions on some implementation details of voxel convolution

### DISCUSSION THREAD

**Discussion by Karbo123** [e3nn/e3nn#300] [Q&A] [answered]
I find the e3nn is a great work in the field of equivariant learning. I recently get interested in the voxel convolution implemented in [v2104](https://github.com/e3nn/e3nn/blob/6fc34a959b/e3nn/nn/models/v2104/voxel_convolution.py), and therefore I may have some questions about it. I would appreciate if the following questions could be addressed.

In the following cases, we only consider steps==(1.0, 1.0, 1.0) for simplicity.
1. I notice that the edge's length is in fact encoded via the function `soft_one_hot_linspace`. However, the number of these lengths are actually limited. For example, if the kernel is of size (3, 3, 3), then their lengths must be chosen from a list of `[0.0, 1.0, sqrt(2), sqrt(3)]`, and other embedded dimensions could be redundant. I wonder if we could embed these lengths in a better way? For example, could we use an 4-dimensional one-hot vector to embed these lengths as 3x3x3 kernal may only have 4 different lengths of its edges (i.e. embedding [1, 0, 0, 0] is for length 0.0, embedding [0, 1, 0, 0] is for length 1.0, and so on)?
2. Related to question#1, I also notice that the maximum length of its embeddings is `r` (see [here](https://github.com/e3nn/e3nn/blob/6fc34a959b/e3nn/nn/models/v2104/voxel_convolution.py#L71)). If the kernel is of size (3, 3, 3), then `r` will be 1.5. But actually the maximum length could be chosen as sqrt(3)=1.732, which is longer than the expected. I think this should be improved by replacing [end=r](https://github.com/e3nn/e3nn/blob/6fc34a959b/e3nn/nn/models/v2104/voxel_convolution.py#L71) with `end=sqrt(x[-1]**2 + y[-1]**2 + z[-1]**2)`. Is that right?

**Comment by mariogeiger** [comment, marked as answer]
These are very good questions.

1. One aim of the radial basis (aka `soft_one_hot_linspace`) is to bias the model toward a smooth radial dependency. So it has been made like that voluntarily to not do it the way you propose. However, these choices has been made based on my intuition taking inspiration from the point cloud version of [TFN](https://arxiv.org/abs/1802.08219), it has to be studied more so I recommend you to make careful comparisons.
2. Only ball shape convolution kernels can lead to an equivariant model. We use cubes because voxels are cubic and `pytorch.nn.conv3d` requires cubic kernels. Therefore we need to set some pixels to zeros. Again, otherwise the model would be not equivariant (I should say "even less equivariant" since the voxel version (compared to the point cloud version) inevitably loses equivariance due to interpolation.).

**Reply by Karbo123** [reply]
Thanks for the quick reply, these help me a lot. However, I encounter another problem when I am trying to find the correct initialization factor for my newly implemented "sparse equivariant convolution". It would be great if some hints are provided.

If point clouds are very sparse, they will only act on the center of the kernel (i.e. the feature is only multipled by the center of the kernel, just like the normal 1x1 conv). Specifically, I assume each entry in the [kernel](https://github.com/e3nn/e3nn/blob/main/e3nn/nn/models/v2104/voxel_convolution.py#L93) will approximately have its second moment to be the same. If this holds, the kernel will equally act on each positions despite different sparsity patterns. Finally outputs should be scaled according to the total number of kernel maps. However, I find that such assumption actually does not hold. For kernel at the center position, its second moment is quite different (much smaller) from other positions, which definitely breaks our assumption. I want to know what is the reason for this, and how to make our assumption correct.

**Reply by Karbo123** [reply]
Here are the minimal codes to reproduce this finding.

``` python
import torch
from e3nn import o3
from e3nn.o3 import FullyConnectedTensorProduct

class Convolution(torch.nn.Module):
    def __init__(self, irreps_in, irreps_out, irreps_sh, kernel_size=(3, 3, 3), **kwargs):
        super().__init__()

        self.irreps_in = o3.Irreps(irreps_in)
        self.irreps_out = o3.Irreps(irreps_out)
        self.irreps_sh = o3.Irreps(irreps_sh)
        self.kwargs = kwargs

        x, y, z = [(torch.arange(k) - k//2).float() for k in kernel_size]
        lattice = torch.stack(torch.meshgrid(x, y, z), dim=-1)

        _, emb_ind = torch.unique(lattice.norm(dim=-1), sorted=True, return_inverse=True)
        self.emb = torch.nn.functional.one_hot(emb_ind).float()
        self.num_radial_basis = self.emb.size(-1)

        self.sh = o3.spherical_harmonics(l=self.irreps_sh, x=lattice, normalize=True,normalization='component')
        self.tp = FullyConnectedTensorProduct(self.irreps_in, self.irreps_sh, self.irreps_out, shared_weights=False)
        self.weight = torch.nn.Parameter(torch.randn(self.num_radial_basis, self.tp.weight_numel))

    def kernel(self):
        weight = self.emb @ self.weight
        kernel = self.tp.right(self.sh, weight)  # [x, y, z, irreps_in.dim, irreps_out.dim]
        kernel = torch.einsum('xyzio->oixyz', kernel)
        return kernel

    def forward(self, x):
        kernel = self.kernel()

        kernel_fla = kernel.reshape(kernel.size(0), kernel.size(1), -1)
        for i in range(kernel_fla.size(-1)): print(i, "==>", (kernel_fla[..., i]**2).mean(), "" if i!=13 else "this is much smaller")
        # we find that kernel_fla[..., 13] is much smaller than others
        # why is this?

        return torch.nn.functional.conv3d(x, kernel, **self.kwargs)



if __name__ == "__main__":
    conv = Convolution("0e + 1e", "0e + 1e + 2e", "0e + 1e + 2e", kernel_size=(3, 3, 3), padding=1)
    x = torch.randn(10, 4, 32, 32, 32)
    y = conv(x)
```

**Reply by mariogeiger** [reply]
It's because spherical harmonics are not properly normalized when evaluated in zero:
```python
sh = conv.sh.reshape(-1, conv.sh.shape[-1])
for i in range(sh.shape[0]):
    print(f"{i} ==> {sh[i].pow(2).mean():.2f}")
# they are all one except for i=13
```

```
sh(0) = [1, 0,0,0,0,0,...]
```

This is the reason why I distinguished the two operations:
- convolutions with neighbors
- self-connection: `self.sc`

**Reply by Karbo123** [reply]
Thank you for this explanation, it is very helpful, and I finally figure out how to correctly normalize the parameters to make outputs satisfy the component normalization assumption.
